Generate serialization of SandboxExtensionHandle

achristensen07 opened the pull request:
#### c31e0dd75e709ee3ffec35dc0ece60754149fdc8
<pre>
Generate serialization of SandboxExtensionHandle
<a href="https://bugs.webkit.org/show_bug.cgi?id=264994">https://bugs.webkit.org/show_bug.cgi?id=264994</a>
<a href="https://rdar.apple.com/118539219">rdar://118539219</a>

Reviewed by Chris Dumez.

All the transformations to get this to work are pretty straightforward,
except the if (dataReference.empty()) check to deserialize a valid, null
SandboxExtensionHandle if sandboxExtensionForType somehow returns an empty
string is replaced by a fast strlen-is-greater-than-0 check in
SandboxExtensionImpl::create to return an empty handle in the creation
process rather than failing to decode and killing the creation process.
This matches existing behavior as close as possible.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtensionImpl::create):
(WebKit::SandboxExtensionImpl::SandboxExtensionImpl):
(WebKit::SandboxExtensionImpl::~SandboxExtensionImpl):
(WebKit::SandboxExtensionImpl::consume):
(WebKit::SandboxExtensionImpl::invalidate):
(WebKit::SandboxExtensionImpl::getSerializedFormat):
(WebKit::SandboxExtensionImpl::sandboxExtensionForType):
(): Deleted.
(WebKit::SandboxExtensionHandle::encode): Deleted.
(WebKit::SandboxExtensionHandle::decode): Deleted.
* Source/WebKit/Shared/SandboxExtension.h:
(WebKit::SandboxExtensionImpl::SandboxExtensionImpl):
(WebKit::SandboxExtensionHandle::SandboxExtensionHandle):
(WebKit::SandboxExtensionHandle::takeImpl):
(WebKit::SandboxExtensionHandle::~SandboxExtensionHandle):
(WebKit::SandboxExtensionHandle::encode): Deleted.
(WebKit::SandboxExtensionHandle::decode): Deleted.
* Source/WebKit/Shared/SandboxExtension.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/270904@main">https://commits.webkit.org/270904@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c91ff0691c913c087b601a1ec5ac3db40b44ca2b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26754 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5369 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27995 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28966 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24462 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7200 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2765 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27016 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4189 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22968 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3682 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3721 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29451 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24403 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24360 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29984 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3759 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1953 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27887 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5208 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4228 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3458 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4111 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
```diff
@@ -41,98 +38,95 @@
+    std::unique_ptr<SandboxExtensionImpl> impl { new SandboxExtensionImpl(path, type, auditToken, flags) };
+    if (!impl->m_token)
+        return nullptr;
+    if (impl->m_token[0] == '\0') // Make sure strlen is > 0 without iterating the whole string.
```
I think we often just do `if (impl->m_token[0])`

achristensen07 replied:
if (!impl->m_token[0])

achristensen07 committed:
Generate serialization of SandboxExtensionHandle
https://bugs.webkit.org/show_bug.cgi?id=264994
rdar://118539219

Reviewed by Chris Dumez.

All the transformations to get this to work are pretty straightforward,
except the if (dataReference.empty()) check to deserialize a valid, null
SandboxExtensionHandle if sandboxExtensionForType somehow returns an empty
string is replaced by a fast strlen-is-greater-than-0 check in
SandboxExtensionImpl::create to return an empty handle in the creation
process rather than failing to decode and killing the creation process.
This matches existing behavior as close as possible.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/SandboxExtensionCocoa.mm:
(WebKit::SandboxExtensionImpl::create):
(WebKit::SandboxExtensionImpl::SandboxExtensionImpl):
(WebKit::SandboxExtensionImpl::~SandboxExtensionImpl):
(WebKit::SandboxExtensionImpl::consume):
(WebKit::SandboxExtensionImpl::invalidate):
(WebKit::SandboxExtensionImpl::getSerializedFormat):
(WebKit::SandboxExtensionImpl::sandboxExtensionForType):
(): Deleted.
(WebKit::SandboxExtensionHandle::encode): Deleted.
(WebKit::SandboxExtensionHandle::decode): Deleted.
* Source/WebKit/Shared/SandboxExtension.h:
(WebKit::SandboxExtensionImpl::SandboxExtensionImpl):
(WebKit::SandboxExtensionHandle::SandboxExtensionHandle):
(WebKit::SandboxExtensionHandle::takeImpl):
(WebKit::SandboxExtensionHandle::~SandboxExtensionHandle):
(WebKit::SandboxExtensionHandle::encode): Deleted.
(WebKit::SandboxExtensionHandle::decode): Deleted.
* Source/WebKit/Shared/SandboxExtension.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/270904@main
